### ISSUE TITLE
ArrayIndexOutOfBoundsException in tryCity method from maxmind

### ISSUE THREAD

**Issue by edelgadoh** [maxmind/GeoIP2-java#180]
Dear maxmind developers,

We are using the last lib 

       <dependency>
            <groupId>com.maxmind.geoip2</groupId>
            <artifactId>geoip2</artifactId>
            <version>2.13.0</version>
        </dependency>

From time to time we observed some errors from maxmind, then you can see the stacktrace below:

`"stack_trace":"java.lang.ArrayIndexOutOfBoundsException: Index 34 out of bounds for length 16\n\tat com.maxmind.db.Decoder$Type.get(Decoder.java:53)\n\tat com.maxmind.db.Decoder.decode(Decoder.java:129)\n\tat com.maxmind.db.Decoder.decode(Decoder.java:88)\n\tat com.maxmind.db.Reader.resolveDataPointer(Reader.java:256)\n\tat com.maxmind.db.Reader.getRecord(Reader.java:176)\n\tat com.maxmind.geoip2.DatabaseReader.get(DatabaseReader.java:271)\n\tat com.maxmind.geoip2.DatabaseReader.tryCity(DatabaseReader.java:334)`

The error start in Decoder.class, specifically in this part

   static enum Type {
        EXTENDED,
        POINTER,
        UTF8_STRING,
        DOUBLE,
        BYTES,
        UINT16,
        UINT32,
        MAP,
        INT32,
        UINT64,
        UINT128,
        ARRAY,
        CONTAINER,
        END_MARKER,
        BOOLEAN,
        FLOAT;

        static final Decoder.Type[] values = values();

        private Type() {
        }

        static Decoder.Type get(int i) {
            return values[i];
        }

sometimes the **i** parameter has the 36 value, but only there are 16 fixed values.

Thanks

**Comment by oschwald** [comment]
It sounds like your database is corrupt.

**Comment by edelgadoh** [comment]
But I works successfully for about 200 requests by second in PRODUCTION, and generate for about 50 errors in 1 minute, why do you say it sounds like it is corrupted?

**Comment by oschwald** [comment]
The fact that it is intermittent supports the idea that it is corrupted. Although the error handling could be improved in this case to throw a nicer exception, what is happening is that it is hitting an unknown and invalid data type in the file while decoding a record. Are you storing the database in a JAR file? If so, please ensure you have [binary filtering disabled](https://github.com/maxmind/GeoIP2-java#packaging-database-in-a-jar).

**Comment by edelgadoh** [comment]
I got it, we are storing the database at shared volume in kubernetes. Yeah, I agree that it could be better a checked exception, for now we are going to try this runtime exception, thank you!

**Comment by edelgadoh** [comment]
@oschwald I have a doubt, is going to be created any task to improve this exception treatment?

**Comment by oschwald** [comment]
Yes, we plan on improving the exception to throw an `InvalidDatabaseException` under this condition. However, you should not be seeing this issue with a valid MMDB file. Under what conditions are you seeing this error?

**Comment by edelgadoh** [comment]
Hi Gregory, great to know that!

As I commented before, we are using a shared volume in kubernetes, and we read the maxmind files from the directory below. We have two jar files that are completely independent and only read from the path to create a bean.
![image](https://user-images.githubusercontent.com/1545116/77251981-534ce600-6c30-11ea-8d52-bd25c60b840a.png)
Another point is that these files are updated from time to time by another team, and we are suspecting about the cache from maxmind, but it was not conclusive yet. 

As I commented before, it's working almost ok for about 98% of each request, and only 2% of the request were generating a runtime exception, but for now we try and catch this runtime exception.

We are going to inspect it in more detail, 
Thank you!

**Comment by oschwald** [comment]
If the files are not replaced atomically on update, you would very likely see errors such as this as well as various `InvalidDatabaseException` errors.

**Comment by edelgadoh** [comment]
@oschwald yeah, the files are replaced atomically.

we reproduced the error in local environment, how to reproduce:
1) make two sets of MindMind files, the first dataset has old files, the second dataset has new files
2) configure JMeter to call the MaxMind API with many ips during 3 minutes
3) while the process in 2 is running, replace between first and second dataset
4) in our tests, we saw errors in logs related to MaxMind database corrupted

In our system, we were using the MaxMind read option: MEMORY_MAP (https://github.com/maxmind/MaxMind-DB-Reader-java#usage) ... after change it to MEMORY option, we couldn't reproduce again this error in our local environment.

**Comment by oschwald** [comment]
Can you share the exact mechanism used for (3) and the OS and file system?
